### ISSUE TITLE
Full Solution Analysis smart toggling

### ISSUE THREAD

**Issue by Therzok** [mono/monodevelop#4873]
Consider adding a smart monitor in the IDE which toggles off full solution analysis in case we're using too much memory or so:
http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/VirtualMemoryNotificationListener.cs,61

We should probably look into how `vm_stat` reports virtual memory and trigger it off automatically and notify the user: https://opensource.apple.com/source/system_cmds/system_cmds-541/vm_stat.tproj/vm_stat.c

> VS bug [#621143](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/621143)

**Comment by Therzok** [comment]
So, we have the following options:

1. use sysctl to get `vm.vm_page_free_target` (optimal value) and check if `vm.page_free_count` is lower.

2. Set a OS dispatch
```
// void (*dispatch_function_t)(void *)

dispatch_function_t *myHandler; // get the data via dispatch_source_get_data
dispatch_source_t source = dispatch_source_create (
	DISPATCH_SOURCE_TYPE_MEMORYPRESSURE,
	0,
	DISPATCH_MEMORYPRESSURE_WARN | DISPATCH_MEMORYPRESSURE_CRITICAL | DISPATCH_MEMORYPRESSURE_NORMAL,
	dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)
);
dispatch_source_set_event_handler (source, myHandler);
dispatch_resume (source);
// We also probably need retain/release semantics, but I think we'll hold the handler while the IDE is alive.
// And cancel when we want to take it off.
// If we hook anything else up (i.e. open resources), we also need to: dispatch_source_set_cancel_handler
```

And when we get to critical, we toggle off full solution analysis. Might be worth surfacing this as an IDE service. This could also help us monitor and flush whatever in-memory caches we have, to reduce memory footprint.

**Comment by mhutch** [comment]
How useful is it to enable Full Solution Analysis, considering we don't surface yet analysis errors through the errors list?

**Comment by Therzok** [comment]
@mhutch todo comments won't happen if full solution analysis is off, for example :p

**Comment by mhutch** [comment]
Ah fair enough. Feels like we're doing a lot of work that won't be surfaced though.

**Comment by Therzok** [comment]
At least, they will only show up for open files.

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-15] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Author: therzok  
 Commit: da75ae971cc414a66af240ab520494afd7938ade (xamarin/monodevelop)
